File: hdf-engine/src/merge.ts
// Merge engine — the TypeScript peer of hdf-engine/go/merge.go (ADR-0016).
// Combines several results documents into one multi-baseline results document:
// one baseline per input baseline renamed `<tool>/<original>`, per-baseline
// provenance in labels, each input's root provenance verbatim under
// extensions["hdf-merge"]. Union semantics — nothing deduplicated, re-keyed or
// dropped. Deterministic for the same inputs in the same order. Kept at
// behavioural parity with Go (see test/merge.test.ts / go/merge_test.go).

import type { HDFResults, EvaluatedBaseline, Component } from '@mitre/hdf-schema';
import { engineVersion } from './version.js';

/** One input to merge: a results document and the name it is recorded under. */
export interface MergeSource {
  name: string;
  doc: HDFResults;
}

export type MergeWarningKind = 'duplicate-baseline-name' | 'label-overwritten';

/** A non-fatal condition merge reports rather than silently resolving. */
export interface MergeWarning {
  kind: MergeWarningKind;
  /** Merged baseline name. */
  name: string;
  /** Merged baseline positions. */
  indices: number[];
  /** Set for label warnings only: the label key that was replaced or removed. */
  label?: string;
}

export interface MergeResult {
  results: HDFResults;
  warnings: MergeWarning[];
}

/** Provenance label keys written on every merged baseline (ADR-0016 §3). */
export const LABEL_TOOL = 'tool';
export const LABEL_TOOL_VERSION = 'toolVersion';
export const LABEL_SOURCE_DOCUMENT = 'sourceDocument';

const MERGE_GENERATOR_NAME = 'hdf-merge';
const MERGE_EXTENSION_KEY = 'hdf-merge';

/**
 * merge combines the sources, in order, into one results document. Returns the
 * document and any warnings (a collision or overwrite is never an error);
 * throws only when there is nothing to merge. Inputs are not mutated.
 */
export function merge(sources: MergeSource[]): MergeResult {
  if (sources.length === 0) {
    throw new Error('merge: no sources');
  }

  const baselines: EvaluatedBaseline[] = [];
  const components: Component[] = [];
  const warnings: MergeWarning[] = [];
  const seenComponent = new Set<string>();
  const provenance: Record<string, unknown>[] = [];
  const positionsByName = new Map<string, number[]>();
  let latest: HDFResults['timestamp'];

  sources.forEach((src, index) => {
    const prefix = toolPrefix(src.doc, index);
    const version = toolVersion(src.doc);

    for (const b of src.doc.baselines ?? []) {
      const name = `${prefix}/${b.name}`;
      const pos = baselines.length;
      const labels: Record<string, string> = { ...(b.labels ?? {}) };

      setLabel(warnings, labels, LABEL_TOOL, prefix, name, pos);
      if (version !== '') {
        setLabel(warnings, labels, LABEL_TOOL_VERSION, version, name, pos);
      } else if (LABEL_TOOL_VERSION in labels) {
        delete labels[LABEL_TOOL_VERSION];
        warnings.push({ kind: 'label-overwritten', name, indices: [pos], label: LABEL_TOOL_VERSION });
      }
      setLabel(warnings, labels, LABEL_SOURCE_DOCUMENT, src.name, name, pos);

      const positions = positionsByName.get(name);
      if (positions) positions.push(pos);
      else positionsByName.set(name, [pos]);

      baselines.push({ ...b, name, labels });
    }

    for (const c of src.doc.components ?? []) {
      if (c.componentId !== undefined) {
        if (seenComponent.has(c.componentId)) continue;
        seenComponent.add(c.componentId);
      }
      components.push(c);
    }

    if (src.doc.timestamp !== undefined && (latest === undefined || millis(src.doc.timestamp) > millis(latest))) {
      latest = src.doc.timestamp;
    }

    provenance.push(sourceProvenance(index, src));
  });

  for (const [name, positions] of positionsByName) {
    if (positions.length > 1) {
      warnings.push({ kind: 'duplicate-baseline-name', name, indices: positions });
    }
  }

  // Key order is fixed so the same inputs serialize identically on every run.
  const results: HDFResults = {
    baselines,
    ...(components.length > 0 ? { components } : {}),
    extensions: { [MERGE_EXTENSION_KEY]: { version: engineVersion, sources: provenance } },
    generator: { name: MERGE_GENERATOR_NAME, version: engineVersion },
    ...(latest !== undefined ? { timestamp: latest } : {}),
  };
  return { results, warnings };
}

function setLabel(
  warnings: MergeWarning[],
  labels: Record<string, string>,
  key: string,
  value: string,
  name: string,
  pos: number,
): void {
  if (key in labels) {
    warnings.push({ kind: 'label-overwritten', name, indices: [pos], label: key });
  }
  labels[key] = value;
}

/** `<tool>` of a merged baseline name: tool.name → generator.name → doc<N>, lower-cased and trimmed. */
function toolPrefix(doc: HDFResults, index: number): string {
  const tool = normalizeName(doc.tool?.name);
  if (tool !== '') return tool;
  const generator = normalizeName(doc.generator?.name);
  if (generator !== '') return generator;
  return `doc${index}`;
}

function normalizeName(s: string | undefined): string {
  return (s ?? '').trim().toLowerCase();
}

/**
 * millis reads a document timestamp for comparison. The schema type is Date,
 * but a document parsed from JSON carries the RFC 3339 string verbatim (the
 * loader does not revive dates), so both are accepted; the value itself is
 * carried through unchanged so output stays byte-faithful to the input.
 */
function millis(ts: Date | string): number {
  return ts instanceof Date ? ts.getTime() : Date.parse(ts);
}

function toolVersion(doc: HDFResults): string {
  return (doc.tool?.version ?? '').trim();
}

/** One input's root metadata, verbatim, for extensions["hdf-merge"].sources[]. */
function sourceProvenance(index: number, src: MergeSource): Record<string, unknown> {
  const entry: Record<string, unknown> = { index, name: src.name };
  if (src.doc.tool !== undefined) entry.tool = src.doc.tool;
  if (src.doc.generator !== undefined) entry.generator = src.doc.generator;
  if (src.doc.timestamp !== undefined) entry.timestamp = src.doc.timestamp;
  if (src.doc.runner !== undefined) entry.runner = src.doc.runner;
  return entry;
}

File: hdf-engine/go/merge.go
// Merge engine — combine several results documents into one multi-baseline
// results document (ADR-0016). One baseline per input baseline, renamed
// `<tool>/<original>`; per-baseline provenance in labels; each input's root
// provenance kept verbatim under extensions["hdf-merge"]. Union semantics: no
// requirement is deduplicated, re-keyed or dropped. Deterministic for the same
// inputs in the same order. Kept at behavioural parity with the TS peer
// (hdf-engine/src/merge.ts); see merge_test.go / test/merge.test.ts.
package hdfengine

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
)

// MergeSource is one input to Merge: a results document and the name it is
// recorded under (its basename, or a caller-supplied label). Name is recorded
// as the sourceDocument label and in the root provenance; it is never a key.
type MergeSource struct {
	Name string
	Doc  hdf.HDFResults
}

// MergeWarningKind classifies a non-fatal condition Merge reports rather than
// silently resolving.
type MergeWarningKind string

const (
	// WarnDuplicateBaselineName: two merged baselines share a name even after
	// the tool prefix. Both are kept; Indices names every position that
	// collides. Read tools key on position (Match.BaselineIndex), so the
	// document stays correct, but name-keyed consumers (hdf diff, baselineRef)
	// cannot tell them apart.
	WarnDuplicateBaselineName MergeWarningKind = "duplicate-baseline-name"
	// WarnLabelOverwritten: a source baseline already carried one of the
	// provenance labels Merge writes (tool, toolVersion, sourceDocument); the
	// source value was replaced — or, for toolVersion when the new source has
	// no tool version, removed, since a stale version would describe the
	// wrong tool. Label names the key, Indices the merged position.
	WarnLabelOverwritten MergeWarningKind = "label-overwritten"
)

// MergeWarning is one reported condition. Name is the merged baseline name;
// Indices are merged baseline positions; Label is set for label warnings only.
type MergeWarning struct {
	Kind    MergeWarningKind `json:"kind"`
	Name    string           `json:"name,omitempty"`
	Indices []int            `json:"indices,omitempty"`
	Label   string           `json:"label,omitempty"`
}

// Provenance label keys written on every merged baseline (ADR-0016 §3).
const (
	LabelTool           = "tool"
	LabelToolVersion    = "toolVersion"
	LabelSourceDocument = "sourceDocument"
)

// mergeGeneratorName is the root generator.name of a merged document.
const mergeGeneratorName = "hdf-merge"

// mergeExtensionKey is the root extensions key carrying per-source provenance.
const mergeExtensionKey = "hdf-merge"

// Merge combines the sources, in order, into one results document. It returns
// the document, any warnings (never an error for a collision or overwrite), and
// an error only when there is nothing to merge. Inputs are not mutated.
func Merge(sources []MergeSource) (hdf.HDFResults, []MergeWarning, error) {
	if len(sources) == 0 {
		return hdf.HDFResults{}, nil, errors.New("merge: no sources")
	}

	out := hdf.HDFResults{
		Generator: &hdf.Generator{Name: mergeGeneratorName, Version: Version()},
	}
	var warnings []MergeWarning
	var latest *time.Time
	seenComponent := map[string]bool{}
	provenance := make([]any, 0, len(sources))
	positionsByName := map[string][]int{}
	var nameOrder []string

	for i := range sources {
		src := sources[i]
		prefix := toolPrefix(src.Doc, i)
		version := toolVersion(src.Doc)

		for j := range src.Doc.Baselines {
			b := src.Doc.Baselines[j]
			nb := b
			nb.Name = prefix + "/" + b.Name
			pos := len(out.Baselines)

			labels := make(map[string]string, len(b.Labels)+3)
			for k, v := range b.Labels {
				labels[k] = v
			}
			warnings = setLabel(warnings, labels, LabelTool, prefix, nb.Name, pos)
			if version != "" {
				warnings = setLabel(warnings, labels, LabelToolVersion, version, nb.Name, pos)
			} else if _, had := labels[LabelToolVersion]; had {
				delete(labels, LabelToolVersion)
				warnings = append(warnings, MergeWarning{Kind: WarnLabelOverwritten, Name: nb.Name, Indices: []int{pos}, Label: LabelToolVersion})
			}
			warnings = setLabel(warnings, labels, LabelSourceDocument, src.Name, nb.Name, pos)
			nb.Labels = labels

			if _, seen := positionsByName[nb.Name]; !seen {
				nameOrder = append(nameOrder, nb.Name)
			}
			positionsByName[nb.Name] = append(positionsByName[nb.Name], pos)
			out.Baselines = append(out.Baselines, nb)
		}

		for j := range src.Doc.Components {
			c := src.Doc.Components[j]
			if c.ComponentID != nil {
				if seenComponent[*c.ComponentID] {
					continue
				}
				seenComponent[*c.ComponentID] = true
			}
			out.Components = append(out.Components, c)
		}

		if ts := src.Doc.Timestamp; ts != nil && (latest == nil || ts.After(*latest)) {
			t := *ts
			latest = &t
		}

		provenance = append(provenance, sourceProvenance(i, src))
	}

	for _, name := range nameOrder {
		if pos := positionsByName[name]; len(pos) > 1 {
			warnings = append(warnings, MergeWarning{Kind: WarnDuplicateBaselineName, Name: name, Indices: pos})
		}
	}

	out.Timestamp = latest
	out.Extensions = map[string]any{
		mergeExtensionKey: map[string]any{
			"version": Version(),
			"sources": provenance,
		},
	}
	return out, warnings, nil
}

// setLabel writes key=value, warning when the baseline already carried the key.
func setLabel(warnings []MergeWarning, labels map[string]string, key, value, name string, pos int) []MergeWarning {
	if _, had := labels[key]; had {
		warnings = append(warnings, MergeWarning{Kind: WarnLabelOverwritten, Name: name, Indices: []int{pos}, Label: key})
	}
	labels[key] = value
	return warnings
}

// toolPrefix is the `<tool>` of a merged baseline name (ADR-0016 §2): the
// root tool.name lower-cased and trimmed; else generator.name the same way;
// else doc<N>.
func toolPrefix(doc hdf.HDFResults, index int) string {
	if doc.Tool != nil && doc.Tool.Name != nil {
		if s := normalizeName(*doc.Tool.Name); s != "" {
			return s
		}
	}
	if doc.Generator != nil {
		if s := normalizeName(doc.Generator.Name); s != "" {
			return s
		}
	}
	return fmt.Sprintf("doc%d", index)
}

func normalizeName(s string) string {
	return strings.ToLower(strings.TrimSpace(s))
}

// toolVersion is the root tool.version, or "" when the source carries none.
func toolVersion(doc hdf.HDFResults) string {
	if doc.Tool != nil && doc.Tool.Version != nil {
		return strings.TrimSpace(*doc.Tool.Version)
	}
	return ""
}

// sourceProvenance is one input's root metadata, verbatim, for
// extensions["hdf-merge"].sources[]: index, name, and — only when the input
// has them — tool, generator, timestamp and runner, each as plain JSON so the
// carrier is the same shape in Go and TS.
func sourceProvenance(index int, src MergeSource) map[string]any {
	entry := map[string]any{"index": index, "name": src.Name}
	if src.Doc.Tool != nil {
		entry["tool"] = asJSONMap(src.Doc.Tool)
	}
	if src.Doc.Generator != nil {
		entry["generator"] = asJSONMap(src.Doc.Generator)
	}
	if src.Doc.Timestamp != nil {
		entry["timestamp"] = src.Doc.Timestamp.Format(time.RFC3339Nano)
	}
	if src.Doc.Runner != nil {
		entry["runner"] = asJSONMap(src.Doc.Runner)
	}
	return entry
}

// asJSONMap round-trips a schema struct through JSON so only the fields it
// actually carries appear (omitempty), exactly as the input serialized them.
func asJSONMap(v any) map[string]any {
	b, err := json.Marshal(v)
	if err != nil {
		return map[string]any{}
	}
	var m map[string]any
	if err := json.Unmarshal(b, &m); err != nil {
		return map[string]any{}
	}
	return m
}
